docs/api/config.md

```diff
@@ -49,6 +49,51 @@ traffic and state size.
 You will also need to specify a directory to write the state data using
 `MESOP_STATE_SESSION_BACKEND_FILE_BASE_DIR`.
 
+#### GCP Firestore
+
+This options uses [GCP Firestore](https://cloud.google.com/firestore?hl=en) to store
+Mesop state sessions. The `(default)` database has a free tier that can be used for
+for small demo applications with low traffic and moderate amounts of state data.
+
+Since Firestore is decoupled from your Mesop server, it allows you to scale vertically
+and horizontally without the considerations you'd need to make for the `memory` and
+`file` backends.
+
+In order to use Firestore, you will need a Google Cloud account with Firestore enabled.
+Follow the instructions for [creating a Firestore in Native mode database](https://cloud.google.com/firestore/docs/create-database-server-client-library#create_a_in_native_mode_database).
+
+Mesop is configured to use the `(default)` Firestore only. The GCP project is determined
+using the Application Default Credentials (ADC) which is automatically configured for
+you on GCP services, such as Cloud Run.
+
+For local development, you can run this command:
+
+```sh
+gcloud auth application-default login
+```
+
+If you have multiple GCP projects, you may need to update the project associated
+with the ADC:
+
+```sh
+GCP_PROJECT=gcp-project
+gcloud config set project $GCP_PROJECT
+gcloud auth application-default set-quota-project $GCP_PROJECT
+```
+
+Mesop leverages Firestore's [TTL policies](https://firebase.google.com/docs/firestore/ttl)
+to delete stale state sessions. This needs to be set up using the following command,
+otherwise old data will accumulate unnecessarily.
+
+```sh
```

Comment:
I'm a little concerned people aren't going to read this far and forget to set this, and then rack up a hefty bill with firestore. In the past when I've used firestore, I never used TTL and it's generally not too bad since you're keying things off a stable id like a user id.

I think it would be good to either: a) print some kind of warning when server starts up and firestore is configured as the state session backend to remind the developer to set TTL or b) detect that TTL policy has _not_ been set and remind the user. I think b) is more work and not sure if it's doable.

Reply:
Yes, that's a great point. I'll add a warning.

The other thing I noticed just now is that using the TTL requires that billing is enabled. The deletion is $0.01 per 100,000 documents.

The other thing is that the TTL deletion takes a while. I have some example records that still haven't been deleted. In the cases where you don't have really large state objects. That's probably fine. Most likely they'd stay under the 1GB daily storage free allowance. But if some's storing a bunch of even 20MB, that could potentially fill up fast.

So one option I think is to clear stale sessions by tracking them per session. So if a process saves a session, it would also save the document ID and timestamp. Then it could try to delete any sessions it has created.

Since the session may be deleted from a different process, we'd have to live with trying to delete sessions that were already deleted. This would lead to extra reads, but should be OK for reasonable demo traffic.

Unfortunately this solution wouldn't be able to delete all stale sessions since the Cloud Run process shutdown with stale sessions still around. So the TTL policy would still be need.

----------

If we want a solution that does not depend on the TTL policy, it can definitely be done. I think the main consideration here is how many extra reads / writes / deletes do we want to incur.

At the moment, I'd vote for enabling requiring the TTL policy.

----------

In terms of billing getting out of control, I think in Firestore, there's a way to set a billing cap. So someone could set it to $5 dollars. But I guess they'd have to be aware of that too.